contracts/utils/RetireToucanCarbon.sol

```diff
@@ -0,0 +1,1405 @@
+            poolRouter[_toucanPool],
+            totalPoolNeeded
+        );
+
+        uint256 unwrappedKLIMA;
+
+        if (_KlimaType == wsKLIMA) {
+            // Get wsKLIMA needed, transfer and unwrap, unstake to KLIMA
+            uint256 wsKLIMANeeded = IwsKLIMA(wsKLIMA).sKLIMATowKLIMA(amountIn);
+
+            IERC20(wsKLIMA).safeTransferFrom(
+                msg.sender,
+                address(this),
+                wsKLIMANeeded
+            );
+            IERC20(wsKLIMA).approve(wsKLIMA, wsKLIMANeeded);
+            unwrappedKLIMA = IwsKLIMA(wsKLIMA).unwrap(wsKLIMANeeded);
+            IERC20(sKLIMA).safeIncreaseAllowance(staking, unwrappedKLIMA);
+            IStaking(staking).unstake(unwrappedKLIMA, false);
+        }
+
+        // If using sKLIMA, transfer in and unstake
+        if (_KlimaType == sKLIMA) {
+            IERC20(sKLIMA).safeTransferFrom(
+                msg.sender,
+                address(this),
+                amountIn
+            );
+            IERC20(sKLIMA).safeIncreaseAllowance(staking, amountIn);
+            IStaking(staking).unstake(amountIn, false);
+        }
+
+        // If using KLIMA, transfer in
+        if (_KlimaType == KLIMA) {
+            IERC20(KLIMA).safeTransferFrom(msg.sender, address(this), amountIn);
+        }
+
+        IERC20(KLIMA).safeIncreaseAllowance(poolRouter[_toucanPool], amountIn);
+
+        uint256[] memory amounts = IUniswapV2Router02(poolRouter[_toucanPool])
+            .swapTokensForExactTokens(
+                totalPoolNeeded,
+                amountIn,
+                path,
+                address(this),
+                block.timestamp
+            );
+        // Return any dust remaining (slippage protection)
+
+        uint256 tradeDust;
+
+        if (_KlimaType == KLIMA) {
+            tradeDust = amountIn - (amounts[0] == 0 ? amounts[1] : amounts[0]);
+            IERC20(KLIMA).safeTransfer(msg.sender, tradeDust);
+        }
+        if (_KlimaType == sKLIMA) {
+            tradeDust = amountIn - (amounts[0] == 0 ? amounts[1] : amounts[0]);
+            IERC20(KLIMA).safeIncreaseAllowance(stakingHelper, tradeDust);
+
+            IStakingHelper(stakingHelper).stake(tradeDust);
+
+            IERC20(sKLIMA).safeTransfer(msg.sender, tradeDust);
+        }
+        if (_KlimaType == wsKLIMA) {
+            tradeDust =
+                unwrappedKLIMA -
+                (amounts[0] == 0 ? amounts[1] : amounts[0]);
+            IERC20(KLIMA).safeIncreaseAllowance(stakingHelper, tradeDust);
+
+            IStakingHelper(stakingHelper).stake(tradeDust);
+            IERC20(sKLIMA).safeIncreaseAllowance(wsKLIMA, tradeDust);
+            uint256 wrappedDust = IwsKLIMA(wsKLIMA).wrap(tradeDust);
+            IERC20(wsKLIMA).safeTransfer(msg.sender, wrappedDust);
+        }
+
+        _retireCarbon(_amount, _beneficiary, _listTCO2, _toucanPool);
+
+        _bondFees(_toucanPool);
+    }
+
+    function retireWithUSDC(
+        uint256 _amount,
```

Comment:
Even though the code is not highly complex, I think clean NatSpec explaining the parameters, e.g. what token's amount is specified with `amount`

Reply:
Should have everything commented/documented more clearly now.